Update docs for slot/3, attr/2, and Phoenix.Component

connorlay opened the pull request:
See https://github.com/phoenixframework/phoenix_live_view/issues/2128

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -8,43 +8,55 @@ defmodule Phoenix.Component do
 
-      defmodule MyComponent do
-        use Phoenix.Component
+  ```elixir
````
Sorry for the nitpick, but can you please keep the Elixir examples four spaces indented instead of using triple backticks?

connorlay replied:
Sure, I'll revert these changes 🙂 

I am curious though if you'd be interested to move towards explicit code blocks for Markdown. Although it is more verbose, it allows tree-sitter and other Markdown parsers to properly handle language injection for accurate syntax highlighting.

chrismccord replied:
In this case, I'm with José on the default 4 space indention being elixir, and ex_doc handles it properly so no need for the extra noise :)

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -70,67 +83,77 @@ defmodule Phoenix.Component do
-        """
-      end
+  ```elixir
+  def show_name(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -70,67 +83,77 @@ defmodule Phoenix.Component do
-      def field_label(assigns) do
-        assigns = assign_new(assigns, :help, fn -> nil end)
+  ```elixir
+  def field_label(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -70,67 +83,77 @@ defmodule Phoenix.Component do
-        """
-      end
+  ```elixir
+  def button(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -145,23 +168,27 @@ defmodule Phoenix.Component do
-        """
-      end
+  ```elixir
+  def unordered_list(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -170,79 +197,84 @@ defmodule Phoenix.Component do
-        """
-      end
+  ```elixir
+  def modal(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -170,79 +197,84 @@ defmodule Phoenix.Component do
+  ```
 
-  ### Named slots with attributes
+  ### Named Slots With Attributes
````
AFAIK we don't use titlecase convention in Phoenix/LiveView.

josevalim commented on lib/phoenix_component.ex:
````diff
@@ -274,24 +308,26 @@ defmodule Phoenix.Component do
-              <% end %>
-            </tr>
+  ```elixir
+  def table(assigns) do
````
Should we add the declarative assigns to this example?

josevalim commented on lib/phoenix_component.ex:
```diff
@@ -660,102 +866,102 @@ defmodule Phoenix.Component do
   the function `@doc` string. To hide a specific attribute, you can set
   the value of `:doc` to `false`.
-
-  ## Examples
```
I think we should keep the examples here too, in case someone jumps to the documentation of the function and wants a quick look without reading all of module docs.

josevalim reviewed (commented):
Thank you @connorlay! I have dropped some comments.

I think the biggest change I would suggest is to introduce and use all declarative assigns early on. So we teach about them as we are teaching components. Then maybe we can have a small section at the end saying they are not necessary but they are recommended. WDYT?

connorlay commented:
> I think the biggest change I would suggest is to introduce and use all declarative assigns early on. So we teach about them as we are teaching components. Then maybe we can have a small section at the end saying they are not necessary but they are recommended. WDYT?

I like that idea! Would you like me to look through the guides as well as part of this PR, or keep this scoped to `Phoenix.Component`? Either way I'll get started on addressing your feedback 👍🏻

josevalim commented:
Just `Phoenix.Component` for now so we keep it focused. :)

josevalim commented on lib/phoenix_component.ex:
```diff
@@ -1,396 +1,435 @@
 defmodule Phoenix.Component do
   @moduledoc ~S'''
-  API for function components.
+  An Elixir API for defining reusable HEEx function components.
```
```suggestion
  Define reusable function components with HEEx templates.
```

jeregrine commented on lib/phoenix_component.ex:
```diff
@@ -573,12 +550,105 @@ defmodule Phoenix.Component do
+  The injected slot docs are formatted as a markdown list:
+
+    * `name` (required) - slot docs. Accepts attributes:
+      * `name` (`:type`) (required) - attr docs. Defaults to `:default`.
```
I think this is supposed to be indented but I'm not 100% sure.

chrismccord replied:
yes indent list items to spaces please

connorlay replied:
This was originally in a code block like this:
```markdown
  * `name` (required) - slot docs. Accepts attributes:
    * `name` (`:type`) (required) - attr docs. Defaults to `:default`.
```

Note that the second item is intended to be rendered as a child of the first, since it is a slot attribute. Does that make sense?

jeregrine replied:
Thats what I thought, it just looked strange and I wanted to gut check. Thanks!

jeregrine commented on lib/phoenix_component.ex:
```diff
@@ -660,100 +735,112 @@ defmodule Phoenix.Component do
 
-  An attribute is declared by its name, type, and options. The following
-  types are supported:
+  * `:any` - any term.
```
we should use the table format here, similar to how Ecto Schema does. 

https://github.com/elixir-ecto/ecto/blob/v3.8.4/lib/ecto/schema.ex#L225

jeregrine reviewed (commented):
Two small comments. One change requested.

chrismccord commented on lib/phoenix_component.ex:
```diff
@@ -1,290 +1,336 @@
+      def project do
+        [
+          # ...
+          compilers: [:phoenix_live_view] ++ Mix.compilers(),
```
Note for the future, the compiler is going away as soon as I can dedicate some time to remove it

chrismccord commented on lib/phoenix_component.ex:
```diff
@@ -1,290 +1,336 @@
-          <% end %>
-        </label>
+  With the `attr/3` macro you have the core ingredients to create reusable function components.
+  But what if you need your function components to support a dynamic number of attributes?
```
```suggestion
  But what if you need your function components to support dynamic attributes, such as common HTML 
  attributes to mix into a component's container?
```
